test/unit/native_deps/native_deps_test.bzl

```diff
@@ -103,11 +103,44 @@ def _bin_has_native_libs_test_impl(ctx):
     _assert_argv_contains(env, action, "-lstatic=native_dep")
     return analysistest.end(env)
 
+def _extract_linker_args(argv):
+    return [a for a in argv if a.startswith("link-arg=-Wl") or a.startswith("link-arg=-l") or a.startswith("-l") or a.endswith(".lo") or a.endswith(".o")]
+
+def _bin_has_native_dep_and_alwayslink_test_impl(ctx):
```

Comment:
Could you also add a test for staticlib and cdylib?

Reply:
That is, a staticlib and a cdylib target that have alwayslink deps?

Reply:
Correct

Reply:
Hm, what does it mean to produce a staticlib with an alwayslink dep? Some objects in deps need to be linked as alwayslink, and we put those into one big archive that staticlib is. We can't mark specific objects in the archive as alwayslink.

It seems we only have 2 options:
1) assume that alwayslink deps are marked alwayslink because there is a circular dependency with some other target that is also covered by the staticlib archive. In this case we don't have to do anything, the circular dependency is taken care of by all objects being in the same archive.
2) assume that alwayslink deps are marked alwayslink because they contain static initialization code that needs to make it into the final artifact. The only immediately available solution I can think of is to mark the whole staticlib as alwayslink. That's going to increase the binary size by a lot.

I propose to go with 1), create an issue describing the problem and add todos referencing the issue, and call it a day.

For cdylib I think our current implementation still makes sense.

Reply:
cdylib test added. I'm not sure I follow the staticlib case, but can you file the bug if you want to track it?